### PR TITLE
Add TSF support

### PR DESCRIPTION
**Pull request by rickywu0421**
Rewriting some code with wrapper macro/function, like replacing
cfg80211_inform_bss_data() with cfg80211_inform_bss(), and so on.
I also add tsf support which provides timing information while
performing cfg80211_ops->scan() routine.
Fix possible buffer overflow by increment the size of every array
to 33, which can tolerate 32 characters (defined by IEEE 802.11
standard) plus one NULL terminator.

### CODE REVIEW

File: vwifi.c [hunk 1]
```diff
@@ -172,24 +172,34 @@ static void inform_dummy_bss(struct owl_context *owl)
...
-        };
-
-        /* array of tags that retrieved from beacon frame or probe responce */
+        struct ieee80211_channel *channel;
+        s32 signal;
+        u64 tsf;
+        /* IE (information element) that retrieved from
+         * beacon frame or probe responce
+         */
+        struct {
```
Comment (jserv) [inline vwifi.c:181]: Don't use anonymous struct unless it is declared with `typedef`.

File: vwifi.c [hunk 2]
```diff
@@ -172,24 +172,34 @@ static void inform_dummy_bss(struct owl_context *owl)
...
-        /* It is posible to use cfg80211_inform_bss() instead. */
-        bss = cfg80211_inform_bss_data(
-            owl->wiphy, &data, CFG80211_BSS_FTYPE_UNKNOWN, ap->bssid, 0,
-            WLAN_CAPABILITY_ESS, 100, ie, ssid_len + 2, GFP_KERNEL);
+        ie = kmalloc(sizeof(*ie) + ssid_len, GFP_KERNEL);
+        ie->eid = WLAN_EID_SSID;
+        ie->len = ssid_len;
+        memcpy(ie->ssid, ap->ssid, ssid_len);
+
+        /* Use CLOCK_BOOTTIME to synchronize with other machines
```
Comment (jserv) [inline vwifi.c:193]: You should explicitly explain why `CLOCK_BOOTTIME` is specified.

**Review by jserv — CHANGES_REQUESTED** [review]
Split the buffer overrun fix into another git commit (and pull request). Minimize the necessary changes.

File: vwifi.c (hunk 1 above)
Comment (rickywu0421) [reply vwifi.c:181]: Ok, should I define this structure outside the function?

Comment (jserv) [reply vwifi.c:181]: No, you just have to name the structure.

**Comment by jserv** [comment]
In this pull request, let's concentrate on TSF support rather than involvement of refactor or overflow fixes.

**Comment by rickywu0421** [comment]
> In this pull request, let's concentrate on TSF support rather than involvement of refactor or overflow fixes.

Ok. Should I move the refactor part to another pull request?

**Comment by jserv** [comment]
> Ok. Should I move the refactor part to another pull request?

Exactly. Always minimize the neccessary changes which would accelerate code review process.

File: vwifi.c [hunk 3]
```diff
@@ -172,24 +172,37 @@ static void inform_dummy_bss(struct owl_context *owl)
...
-        };
-
-        /* array of tags that retrieved from beacon frame or probe responce */
+        struct ieee80211_channel *channel;
+        s32 signal;
+        u64 tsf;
+        /* IE (information element) that retrieved from
+         * beacon frame or probe responce.
+         */
+        struct ie_struct {
```
Comment (jserv) [inline vwifi.c:181]: Rename `ie_struct` to `__ie` or `ie_s`.

File: vwifi.c [hunk 4]
```diff
@@ -172,24 +172,37 @@ static void inform_dummy_bss(struct owl_context *owl)
...
+            u8 eid, len;
+            u8 ssid[]; /* zero length array */
+        } * ie;
```
Comment (jserv) [inline vwifi.c:184]: Move the declaration of `ie` object in the next line.

File: vwifi.c [hunk 5]
```diff
@@ -322,8 +335,11 @@ static int owl_scan(struct wiphy *wiphy, struct cfg80211_scan_request *request)
 
     mutex_unlock(&owl->lock);
 
+    inform_dummy_bss(owl);
+
     if (!schedule_work(&owl->ws_scan))
         return -EBUSY;
+
```
Comment (jserv) [inline vwifi.c:342]: Don't change this line.

**Comment by jserv** [comment]
Can you provide test item(s) in directory `scripts` for timing synchronization function? e.g. check the results of command `iw dev scan`.

**Comment by rickywu0421** [comment]
I add a timing test in scripts/verify.sh. I compare the value of tsf and uptime, if the difference is higher than 500ms, then the test fails. I don't know if the threshold value "500ms" is reasonable.

File: scripts/verify.sh [hunk 6]
```diff
@@ -16,20 +16,35 @@ if [ $? -ne 0 ]; then
...
+    cat scan_result.log | grep -o -E '([[:xdigit:]]{1,2}:){5}[[:xdigit:]]{1,2}'| head -n 1 > scan_bssid.log
     sudo iw dev owl0 connect MyHomeWiFi
     sudo iw dev owl0 link | grep -o -E '([[:xdigit:]]{1,2}:){5}[[:xdigit:]]{1,2}' > connected.log
 
     DIFF=$(diff connected.log scan_bssid.log)
     if [ "$DIFF" != "" ]; then
         final_ret=3
     fi
+
+    # verify tsf (in usec)
```
Comment (jserv) [inline scripts/verify.sh:33]: Use uppercase `TSF` to indicate the functionality of WiFi.

**Review by jserv — CHANGES_REQUESTED** [review]
Describe TSF within the git commit message(s), so that more developers can be aware of this.

**Comment by jserv** [comment]
Thank @rickywu0421 for contributing!

**Pull request merged**